poly-graph-persistent/src/Data/Graph/HGraph/Persistent/TH.hs

```diff
@@ -0,0 +1,132 @@
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+module Data.Graph.HGraph.Persistent.TH
+  ( UniquenessCheck(..)
+  , mkUniquenessChecks
+  , mkUniquenessCheck
+  ) where
+
+import Control.Arrow ((&&&))
+import Data.Char (toLower, toUpper)
+import Data.List.NonEmpty (nonEmpty)
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import Data.Maybe (mapMaybe)
+import Data.Monoid ((<>))
+import Data.Text (Text, unpack, cons, uncons)
+import Database.Persist
+import Database.Persist.Quasi (nullable)
+import Database.Persist.TH
+import Language.Haskell.TH
+
+class UniquenessCheck a where
+  couldCauseUniquenessViolation :: a -> a -> Bool
```

Comment:
I'm happy to change it, but it is meant to reflect that this is conservative. Suppose we have a constraint on a foreign key and an int:
```haskell
SomeUnique otherId count
```

Because all the `otherId`s aren't filled in yet, we have to assume the worst case - all of these entities might point at the same entity using `otherId`. Therefore, we can only compare `count`. `couldCauseUniquenessViolation` could return `True` when inserting the given entity would be safe since we can't know with certainty.